c++/src/capnp/cc_capnp_library.bzl

```diff
@@ -0,0 +1,114 @@
+"""Bazel rule to compile .capnp files into c++."""
+
+capnp_provider = provider("Capnproto Provider", fields = {
+    "includes": "includes for this target (transitive)",
+    "inputs": "src + data for the target",
+    "src_prefix": "src_prefix of the target",
+})
+
+def _workspace_path(label, path):
+    if label.workspace_root == "":
+        return path
+    return label.workspace_root + "/" + path
+
+def _impl(ctx):
+    label = ctx.label
+    src_prefix = _workspace_path(label, ctx.attr.src_prefix)
+    includes = []
+
+    inputs = ctx.files.srcs + ctx.files.data
+    for dep_target in ctx.attr.deps:
+        includes += dep_target[capnp_provider].includes
+        inputs += dep_target[capnp_provider].inputs
+
+    if src_prefix != "":
+        includes.append(src_prefix)
+
+    system_include = _workspace_path(ctx.attr._capnp_system.label, "src")
+
+    out_dir = ctx.var["GENDIR"]
+    if src_prefix != "":
+        out_dir = out_dir + "/" + src_prefix
+
+    cc_out = "-o%s:%s" % (ctx.executable._capnpc_cxx.path, out_dir)
+    args = ["compile", "--verbose", cc_out] + \
+           ["-I" + inc for inc in includes] + \
+           ["-I" + system_include]
+    if src_prefix != "":
+        args += ["--src-prefix", src_prefix]
+
+    args += [s.path for s in ctx.files.srcs]
+
+    ctx.actions.run(
+        inputs = inputs + ctx.files._capnpc_cxx + ctx.files._capnpc_capnp + ctx.files._capnp_system,
+        outputs = ctx.outputs.outs,
+        executable = ctx.executable._capnpc,
+        arguments = args,
+        mnemonic = "GenCapnp",
+    )
+
+    return [
+        capnp_provider(
+            includes = includes,
+            inputs = inputs,
+            src_prefix = src_prefix,
+        ),
+    ]
+
+_capnp_gen = rule(
+    attrs = {
+        "srcs": attr.label_list(allow_files = True),
+        "deps": attr.label_list(providers = [capnp_provider]),
+        "data": attr.label_list(allow_files = True),
+        "outs": attr.output_list(),
+        "src_prefix": attr.string(),
+        "_capnpc": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnp_tool"),
+        "_capnpc_cxx": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnpc-c++"),
+        "_capnpc_capnp": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnpc-capnp"),
```

Comment:
maybe use toolchains instead?

Reply:
Frankly, after setting up c++ toolchain it seems like way too much work for very little benefit.

Reply:
The c++ toolchain stuff is a very different and tedious beast compared to regular toolchains. A capnp toolchain would be very simple and very beneficial. I can make a mock-up tomorrow if you like.

Reply:
I gave it a bit of thought and I don't think toolchain is necessary in this case. These files build capnp from these sources and use it. The goal is not to provide a separate independent implementation of cc_capnp_library. I don't think we need a toolchain complexity here.